File: CONTROLS/RunFD_Diffcontrols.m

%% Example with data chnaged programmatically in Matlab
clear;close all;

%% Import the OpenSim modeling classes
import org.opensim.modeling.*

%% Excel file
xlsfile='/Users/dc547/Desktop/Work/Teaching/HL40064 - MSK Modelling/2020_21/Projects/Kicking DATA/EMGs.xlsx'; %chnage it accordingly 
osimModel_file='/Users/dc547/Desktop/Work/Teaching/HL40064 - MSK Modelling/2020_21/Projects/Kicking DATA/SoccerKick/SoccerKickingModel.osim';
%Read excel file
[num,txt,~] = xlsread(xlsfile) ;
[row,col]=size(num);
time=num(:,1);

n_sims=50;
% create a matrix of zeros of the same num of col and rows as the one
% inputted - this is created n_sims times.

base_EMGs=zeros(row,col-1,n_sims); %this is a 3D matrix!

% change magnitude
for j=1:n_sims %running 50 times indeed
    
  dec_factor=j/100;
    
  activation_coeff=[1 1 1 1 1 1 1 1]*dec_factor;  % [rect_fem_r	bifemlh_r	bifemsh_r	vast_int_r	glut_max_r	psoas_r	tib_ant_r	med_gas_r]
  base_EMGs(:,:,j)=num(:,2:end).*activation_coeff;
    
  create_control_file( base_EMGs(:,:,j),'controls_kick.xml',[ 'new_control_' num2str(2) '.xml'],txt) ;

  
%% Run FD
  


%% LOAD states
% % states = load_sto_file(states_file);
f_time=time(end);
%% Read osim model
osimModel = Model(osimModel_file); %add paths for the osim model file

ft=ForwardTool();
ft.setModel(osimModel);
ft.setSolveForEquilibrium(1);
ft.setErrorTolerance(0.0000001);
ft.setMaxDT(time(2,1)/100);
ft.setMinDT(time(2,1)/1000);
ft.setInitialTime(0);
ft.setFinalTime(f_time);
ft.setName([ '/Users/dc547/Desktop/Work/Teaching/HL40064 - MSK Modelling/2020_21/Projects/Kicking DATA/new_res_' num2str(2) '.sto']);%names to be changed depending on input values
% % ft.setStatesFileName(states_file);
ft.run();

  
  
end

File: CONTROLS/RunFD_Diffcontrols.py
"""Example with data changed programmatically in Python.

Runs multiple forward dynamics simulations with varying activation levels.

Originally a MATLAB script by D. Cazzola, Uni of Bath
Python version
"""

import numpy as np
import pandas as pd

from CONTROLS.create_control_file import create_control_file

# Requires: import opensim  (uncomment when opensim Python bindings are available)


def main():
    # File paths (update to your local paths)
    xlsfile = 'EMGs.xlsx'
    osimModel_file = 'SoccerKickingModel.osim'

    # Read excel file
    df = pd.read_excel(xlsfile)
    txt = list(df.columns)
    num = df.values
    row, col = num.shape
    time = num[:, 0]

    n_sims = 50

    # Create a 3D array of zeros: (row, col-1, n_sims)
    base_EMGs = np.zeros((row, col - 1, n_sims))

    for j in range(n_sims):
        dec_factor = (j + 1) / 100.0

        activation_coeff = np.ones(col - 1) * dec_factor
        base_EMGs[:, :, j] = num[:, 1:] * activation_coeff

        # Build control data with time as first column
        control_data = np.column_stack([time, base_EMGs[:, :, j]])

        create_control_file(
            control_data,
            'controls_kick.xml',
            f'new_control_{j + 1}.xml',
            txt
        )

        # --- Run Forward Dynamics ---
        # Uncomment the following when opensim Python bindings are available:
        #
        # import opensim
        # f_time = time[-1]
        # osimModel = opensim.Model(osimModel_file)
        #
        # ft = opensim.ForwardTool()
        # ft.setModel(osimModel)
        # ft.setSolveForEquilibrium(True)
        # ft.setErrorTolerance(1e-7)
        # ft.setMaxDT(time[1] / 100.0)
        # ft.setMinDT(time[1] / 1000.0)
        # ft.setInitialTime(0)
        # ft.setFinalTime(f_time)
        # ft.setName(f'new_res_{j + 1}.sto')
        # ft.run()

        print(f'Simulation {j + 1}/{n_sims} complete (dec_factor={dec_factor:.2f})')


if __name__ == '__main__':
    main()
